src/jose/jws/general/verifier.ts

```diff
@@ -0,0 +1,112 @@
+import type { GeneralJws, Signature } from './types';
+import type { JwkPublic, VerifyFn } from '../../types';
+import type { VerificationMethod } from '../../../did/did-resolver';
+
+import { base64url } from 'multiformats/bases/base64';
+import { DIDResolver } from '../../../did/did-resolver';
+import { verify as verifyEd25519 } from '../../algorithms/ed25519';
+import { verify as verifySecp256k1 } from '../../algorithms/secp256k1';
+
+const verifiers: { [key:string]: VerifyFn } = {
+  'Ed25519'   : verifyEd25519,
+  'secp256k1' : verifySecp256k1
+};
+
+type VerificationResult = {
+  signers: string[];
+};
+
+// TODO: add logic to prevent validating duplicate signatures
+export class GeneralJwsVerifier {
+  jws: GeneralJws;
+
+  constructor(jws: GeneralJws) {
+    this.jws = jws;
+  }
+
+  async verify(didResolver: DIDResolver): Promise<VerificationResult> {
+    const signers: string[] = [];
+
+    for (let signature of this.jws.signatures) {
+      const protectedBytes = base64url.baseDecode(signature.protected);
+      const protectedJson = new TextDecoder().decode(protectedBytes);
+
+      const { kid } = JSON.parse(protectedJson);
+      const did = this.extractDid(kid);
+      const publicKeyJwk = await this.getPublicKey(did, kid, didResolver);
+
+      const isVerified = await this.verifySignature(this.jws.payload, signature, publicKeyJwk);
+
+      if (isVerified) {
+        signers.push(did);
+      } else {
+        throw new Error('signature verification failed');
```

Comment:
Generic question on error messages. Is it your intention that we'll revisit each of these messages in a separate pass? Or we should optimistically improve them? Kinda with the opinion that some error messages need more contexts. For instance here would be nice to know which DID failed validation.

Reply:
100% that some error messages need more context. I think we can incrementally improve them as we run into them. went ahead and took care of this one: https://github.com/TBD54566975/dwn-sdk-js/commit/ce65f493f63bf276b01337fed8b78fa1187ba8ba